lib/SIL/SILFunctionType.cpp

```diff
@@ -323,6 +323,81 @@ CanSILFunctionType SILFunctionType::getAutoDiffDerivativeFunctionType(
                               getWitnessMethodConformanceOrNone());
 }
 
+CanSILFunctionType SILFunctionType::getAutoDiffTransposeFunctionType(
+    IndexSubset *parameterIndices, Lowering::TypeConverter &TC,
+    LookupConformanceFn lookupConformance, CanGenericSignature genSig) {
+  // Get the canonical derivative function generic signature.
+  if (!genSig)
+    genSig = getGenericSignature();
+  genSig = getAutoDiffDerivativeFunctionGenericSignature(
+      genSig, getParameters(), parameterIndices, &TC.M);
+  Lowering::GenericContextScope genericContextScope(TC, genSig);
+
+  // Given a type, returns its formal SIL parameter info.
+  auto getParameterInfoForOriginalResult = [&](
```

Comment:
I might recommend moving `getParameterInfoForOriginalResult` and `getResultInfoForOriginalParameter` to the top level for deduplication but it's NBD because they'll be removed in https://github.com/apple/swift/pull/26709.